File: models/mlb/cartridges/MLB-RP36/verify_run.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import hashlib
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

from run_lock import ROOT, RUN_ROOT, aggregate_hash, build_snapshot, file_hash, input_inventory, stable_json


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Verify an MLB-RP36 reliever addendum run lock.")
    parser.add_argument("--date", required=True, help="Slate date, YYYY-MM-DD.")
    return parser.parse_args()


def read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def sha256_text(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def first_diff(left: Any, right: Any, prefix: str = "$") -> str | None:
    if left == right:
        return None
    if type(left) is not type(right):
        return f"{prefix}: type {type(left).__name__} != {type(right).__name__}"
    if not isinstance(left, (dict, list)):
        return f"{prefix}: {left!r} != {right!r}"
    if isinstance(left, list):
        if len(left) != len(right):
            return f"{prefix}: length {len(left)} != {len(right)}"
        for index, (left_item, right_item) in enumerate(zip(left, right)):
            nested = first_diff(left_item, right_item, f"{prefix}[{index}]")
            if nested:
                return nested
        return None
    keys = sorted(set(left) | set(right))
    for key in keys:
        if key not in left:
            return f"{prefix}.{key}: missing on actual"
        if key not in right:
            return f"{prefix}.{key}: missing on expected"
        nested = first_diff(left[key], right[key], f"{prefix}.{key}")
        if nested:
            return nested
    return None


def assert_file_lock(lock_path: Path, rows_key: str, expected_hash: str, label: str, exclude_roles: set[str] | None = None) -> int:
    exclude_roles = exclude_roles or set()
    lock = read_json(lock_path)
    actual_rows = [{**entry, **file_hash(entry["path"])} for entry in lock.get(rows_key, [])]
    for expected, actual in zip(lock.get(rows_key, []), actual_rows):
        if bool(expected.get("exists")) != bool(actual.get("exists")) or expected.get("sha256") != actual.get("sha256"):
            raise RuntimeError(f"{label} drift: {actual.get('path')}")
    hash_rows = [row for row in actual_rows if row.get("role") not in exclude_roles]
    actual_hash = aggregate_hash(hash_rows)
    if actual_hash != expected_hash:
        raise RuntimeError(f"{label} hash mismatch: expected {expected_hash}, got {actual_hash}")
    return len(actual_rows)


def assert_input_lock(lock_path: Path, expected_hash: str, date: str) -> int:
    lock = read_json(lock_path)
    actual_rows = input_inventory(date)
    diff = first_diff(actual_rows, lock.get("inputs", []))
    if diff:
        raise RuntimeError(f"Input lock drift: {diff}")
    actual_hash = sha256_text(stable_json(actual_rows))
    if actual_hash != expected_hash:
        raise RuntimeError(f"Input hash mismatch: expected {expected_hash}, got {actual_hash}")
    return len(actual_rows)


def verify_snapshot(date: str) -> None:
    completed = subprocess.run(
        [sys.executable, str(ROOT / "models" / "mlb" / "cartridges" / "MLB-RP36" / "verify_snapshot.py"), "--date", date],
        cwd=ROOT,
        text=True,
        capture_output=True,
        check=False,
    )
    if completed.returncode:
        raise RuntimeError((completed.stdout + completed.stderr).strip())


def main() -> int:
    args = parse_args()
    date = args.date
    run_dir = RUN_ROOT / date
    run = read_json(run_dir / "run.json")
    snapshot = read_json(run_dir / "snapshot.json")
    actual_snapshot = build_snapshot(date)
    diff = first_diff(actual_snapshot, snapshot)
    if diff:
        raise RuntimeError(f"MLB-RP36 snapshot mismatch: {diff}")

    source_files = assert_file_lock(run_dir / "files.lock.json", "files", run["sourceHash"], "Source lock")
    inputs = assert_input_lock(run_dir / "inputs.lock.json", run["inputHash"], date)
    outputs = assert_file_lock(
        run_dir / "outputs.lock.json",
        "outputs",
        run["outputHash"],
        "Output lock",
        exclude_roles={"run-manifest"},
    )
    verify_snapshot(date)

    print(json.dumps({
        "runId": run["runId"],
        "status": "verified",
        "sourceFiles": source_files,
        "inputs": inputs,
        "outputs": outputs,
        "artifactHash": run["artifactHash"],
        "artifactSummary": run["artifactSummary"],
    }, indent=2))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(error, file=sys.stderr)
        raise SystemExit(1)

File: models/mlb/cartridges/MLB-M1/verify_run.mjs
import fs from 'node:fs/promises'
import path from 'node:path'
import crypto from 'node:crypto'
import { buildM1Snapshot } from './snapshot.mjs'

const rootDir = path.resolve(import.meta.dirname, '..', '..', '..', '..')

const parseArgs = () => {
  const args = process.argv.slice(2)
  const options = { date: '' }
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === '--date') {
      options.date = args[index + 1]
      index += 1
    }
  }
  if (!/^\d{4}-\d{2}-\d{2}$/.test(options.date)) throw new Error('Pass --date YYYY-MM-DD')
  return options
}

const readJson = async (relativePath) => JSON.parse(await fs.readFile(path.resolve(rootDir, relativePath), 'utf8'))

const writeJson = async (relativePath, payload) => {
  const absolute = path.resolve(rootDir, relativePath)
  await fs.mkdir(path.dirname(absolute), { recursive: true })
  await fs.writeFile(absolute, `${JSON.stringify(payload, null, 2)}\n`)
}

const stableJson = (value) => {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(',')}]`
  if (value && typeof value === 'object') {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${stableJson(value[key])}`).join(',')}}`
  }
  return JSON.stringify(value)
}

const sha256Text = (text) => crypto.createHash('sha256').update(text).digest('hex')

const fileHash = async (filePath) => {
  const absolute = path.resolve(rootDir, filePath)
  try {
    const bytes = await fs.readFile(absolute)
    return {
      path: filePath,
      exists: true,
      sha256: crypto.createHash('sha256').update(bytes).digest('hex')
    }
  } catch (error) {
    if (error.code !== 'ENOENT') throw error
    return {
      path: filePath,
      exists: false,
      sha256: null
    }
  }
}

const aggregateHash = (entries) => sha256Text(stableJson(
  (entries || []).map((entry) => ({
    path: entry.path,
    role: entry.role || null,
    exists: Boolean(entry.exists),
    sha256: entry.sha256 || null
  })).sort((left, right) => String(left.path).localeCompare(String(right.path)))
))

const firstDiff = (left, right, prefix = '$') => {
  if (Object.is(left, right)) return null
  if (typeof left !== typeof right) return `${prefix}: type ${typeof left} !== ${typeof right}`
  if (left === null || right === null || typeof left !== 'object') return `${prefix}: ${JSON.stringify(left)} !== ${JSON.stringify(right)}`
  if (Array.isArray(left) !== Array.isArray(right)) return `${prefix}: array mismatch`
  if (Array.isArray(left)) {
    if (left.length !== right.length) return `${prefix}: length ${left.length} !== ${right.length}`
    for (let index = 0; index < left.length; index += 1) {
      const nested = firstDiff(left[index], right[index], `${prefix}[${index}]`)
      if (nested) return nested
    }
    return null
  }
  const keys = Array.from(new Set([...Object.keys(left), ...Object.keys(right)])).sort()
  for (const key of keys) {
    if (!(key in left)) return `${prefix}.${key}: missing on actual`
    if (!(key in right)) return `${prefix}.${key}: missing on expected`
    const nested = firstDiff(left[key], right[key], `${prefix}.${key}`)
    if (nested) return nested
  }
  return null
}

const assertLock = async ({ lockPath, rowsKey, expectedHash, label, excludeRolesFromHash = new Set() }) => {
  const lock = await readJson(lockPath)
  const rows = await Promise.all((lock[rowsKey] || []).map(async (entry) => ({
    ...entry,
    ...(await fileHash(entry.path))
  })))
  const mismatches = rows.filter((row, index) => {
    const expected = lock[rowsKey][index]
    return Boolean(expected.exists) !== row.exists || (expected.sha256 || null) !== (row.sha256 || null)
  })
  if (mismatches.length) {
    const first = mismatches[0]
    throw new Error(`${label} drift: ${first.path}`)
  }
  const hashRows = rows.filter((row) => !excludeRolesFromHash.has(row.role))
  const hash = aggregateHash(hashRows)
  if (hash !== expectedHash) throw new Error(`${label} hash mismatch: expected ${expectedHash}, got ${hash}`)
  return rows.length
}

const main = async () => {
  const { date } = parseArgs()
  const runDir = `data-private/model-runs/mlb/MLB-M1/${date}`
  const run = await readJson(`${runDir}/run.json`)
  const snapshot = await readJson(`${runDir}/snapshot.json`)
  const actualSnapshot = await buildM1Snapshot({ date })
  const diff = firstDiff(actualSnapshot, snapshot)
  if (diff) {
    const actualPath = `${runDir}/snapshot.actual.json`
    await writeJson(actualPath, actualSnapshot)
    throw new Error(`MLB-M1 run snapshot mismatch: ${diff}. Actual written to ${actualPath}`)
  }
  const sourceFiles = await assertLock({
    lockPath: `${runDir}/files.lock.json`,
    rowsKey: 'files',
    expectedHash: run.sourceHash,
    label: 'Source lock'
  })
  const inputs = await assertLock({
    lockPath: `${runDir}/inputs.lock.json`,
    rowsKey: 'inputs',
    expectedHash: run.inputHash,
    label: 'Input lock'
  })
  const outputs = await assertLock({
    lockPath: `${runDir}/outputs.lock.json`,
    rowsKey: 'outputs',
    expectedHash: run.outputHash,
    label: 'Output lock',
    excludeRolesFromHash: new Set(['run-manifest'])
  })
  console.log(JSON.stringify({
    runId: run.runId,
    status: 'verified',
    sourceFiles,
    inputs,
    outputs,
    snapshotHash: run.snapshotHash,
    artifactSummary: run.artifactSummary
  }, null, 2))
}

main().catch((error) => {
  console.error(error.message || error)
  process.exitCode = 1
})
